BubbleScan-AI/Bubble_Scan_AI.py

```diff
@@ -0,0 +1,413 @@
+            return None
+        # The filled_bubble_index corresponds to the digit
+        return filled_bubble_index
+
+    def student_id(self, roi, num_columns=10, num_bubbles=10):
+        student_id = ''
+
+        # Calculating the width of each response column in the ROI
+        digit_width = roi.shape[1] // num_columns
+
+        # Processing each column to identify the filled bubble
+        for i in range(num_columns):
+            column_roi = roi[:, i * digit_width:(i + 1) * digit_width]
+
+            # Column ROI to grayscale and threshold it to create a binary image
+            gray = cv2.cvtColor(column_roi, cv2.COLOR_BGR2GRAY)
+            _, binary = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY_INV)
+
+            # Column to determine the filled bubble
+            filled_digit = self.bubble_column(binary, num_bubbles)
+
+            # Appending the filled digit to the student ID, or 'X' if not detected
+            student_id += str(filled_digit) if filled_digit is not None else 'X'
+
+        return student_id
+
+    def extract_responses(self):
+        students_results = []
+        base_folder_path = os.path.join(self.source_folder, "ROIs")
+
+        student_image_folders = sorted(os.listdir(base_folder_path),
+                                       key=lambda x: [int(c) if c.isdigit() else c for c in re.split(r'(\d+)', x)])
+        for student_image_folder in student_image_folders:
+            student_folder_path = os.path.join(base_folder_path, student_image_folder)
+            if os.path.isdir(student_folder_path):
+
+                first_column_path = os.path.join(student_folder_path, 'first_column_roi.jpg')
+                second_column_path = os.path.join(student_folder_path, 'second_column_roi.jpg')
+                student_id_path = os.path.join(student_folder_path, 'student_id_roi.jpg')
+
+                if not all([os.path.exists(first_column_path), os.path.exists(second_column_path),
+                            os.path.exists(student_id_path)]):
+                    print(f"Missing ROI images in folder: {student_folder_path}")
+                    continue
+
+                # Images in the order they are found
+                student_id = self.student_id(cv2.imread(student_id_path))
+                responses_first = self.roi(cv2.imread(first_column_path), start_question_num=1)
+                responses_second = self.roi(cv2.imread(second_column_path), start_question_num=26)
+
+                student_data = {
+                    "studentID": student_id,
+                    "answers": {**responses_first, **responses_second}
+                }
+                students_results.append(student_data)
+
+        final_output = {"students": students_results}
+
+        with open('result_data.json', 'w') as json_file:
+            json.dump(final_output, json_file, indent=4)
+
+        print("Processing complete. Data saved to 'result_data.json'.")
```

Comment:
It would probably make more sense for the extract_responses function to return the json data, rather than write it to a file.

Reply:
Yes, that's the idea. AI team and SD team are planning to work together for integration. We will discuss and alter the output as required. This is a whole working module to run and test the output.